### PR TITLE
Uploading starter assets to Applab

### PR DESCRIPTION
**Pull request by maddiedierker**
Follow-up to #30218 
[LP-609](https://codedotorg.atlassian.net/browse/LP-609)
[Technical design](https://docs.google.com/document/d/1KrAAv4klJhLL_siQo-t6F-2zJrCOOub6Z7W2b9IqDto/edit#heading=h.gbynda1ya0gp)

### What it does
- **Allows levelbuilders to upload new starter assets when in levelbuilder mode.** This will allow devs to upload images locally without affecting other environments -- as long as they do not commit the changes to that level's .level file. (Addresses @islemaster's [comment](https://github.com/code-dot-org/code-dot-org/pull/30218#pullrequestreview-276200870).)
- Detour: Rename 'guid' to 'uuid' everywhere to avoid confusion.

### CODE REVIEW

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 1]
```diff
@@ -28,20 +18,71 @@ def show
...
+    uuid_name = @level.starter_assets[friendly_name]
+    file_obj = get_object(uuid_name)
+    content_type = file_content_type(File.extname(uuid_name))
 
     expires_in 1.hour, public: true
     send_data read_file(file_obj), type: content_type, disposition: 'inline'
   end
 
+  # POST /level_starter_assets/:level_name
+  def upload
```
Comment (davidsbailey) [inline dashboard/app/controllers/level_starter_assets_controller.rb:33]: I think the standard name for this would be `create`, and the standard name for `file` would be `get`.

File: dashboard/config/routes.rb [hunk 2]
```diff
@@ -234,6 +234,7 @@
   resources :level_starter_assets, only: [:show], param: 'level_name' do
     member do
       get '/:filename', to: 'level_starter_assets#file'
+      post '', to: 'level_starter_assets#upload'
```
Comment (davidsbailey) [inline dashboard/config/routes.rb:237]: if you rename your controller methods to `get` and `create`, you should be able to remove the `member do` block and just list these in `resources :level_starter_assets, only: [:show, :create]`. For example, the following resource definition:
```
  concern :section_api_routes do
    resources :sections, only: [:index, :show, :create, ...]
```
generates the following routes:
```
$ rails routes | grep /api/v1/sections
...
api_v1_sections GET      /api/v1/sections(.:format)                                                                   api/v1/sections#index
                POST     /api/v1/sections(.:format)                                                                   api/v1/sections#create
                GET      /api/v1/sections/:id(.:format)                                                               api/v1/sections#show
```

I may not have the syntax exactly right in my code snippet at the top, but it's worth figuring out the rails syntax / convention for this, and saving the `member` directive for when you really need to do something special.

**Review by davidsbailey — COMMENTED** [review]
a couple initial comments...

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 3]
```diff
@@ -28,19 +18,60 @@ def show
...
+    return head :forbidden unless current_user&.levelbuilder? && Rails.application.config.levelbuilder_mode
```
Comment (davidsbailey) [inline dashboard/app/controllers/level_starter_assets_controller.rb]: this will work, but it for "extra credit" it would be ideal to (1) move the user check into an ability following this example: https://github.com/code-dot-org/code-dot-org/blob/cffd6d7ce7344ad3d24fc49f3c34d4c2bb0c0d59/dashboard/app/models/ability.rb#L163-L164
and (2) use `before_action :require_levelbuilder_mode, except: :show` instead of the `Rails.application.config` check.

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 4]
```diff
@@ -28,19 +18,60 @@ def show
...
+
+    # Client expects a single file upload, so raise an error if params[:files] contains more than one file.
+    if params[:files].length > 1
+      raise "One file upload expected. Actual: #{params[:files].length}"
+    end
+
+    upload = params[:files]&.first
+    friendly_name = upload.original_filename
+    # Replace the friendly file name with a UUID for storage in S3 to avoid naming conflicts.
+    uuid_name = SecureRandom.uuid + File.extname(friendly_name)
```
Comment (davidsbailey) [inline dashboard/app/controllers/level_starter_assets_controller.rb]: we should probably use a whitelist of allowed filename extensions, in part to keep any characters which are not [safe characters](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-keys) out of the S3 key. This seems like a good list to copy: https://github.com/code-dot-org/code-dot-org/blob/ea5ee2cecca58c59fc0f609fc78cf63d523059c5/shared/middleware/helpers/asset_bucket.rb#L9-L11

File: dashboard/app/models/levels/level.rb [hunk 5]
```diff
@@ -602,6 +602,19 @@ def age_13_required?
...
   end
 
+  # Add a starter asset to the level and save it in properties.
+  # Starter assets are stored as an object, where the key is the
+  # friendly filename and the value is the UUID filename stored in S3:
+  # {
+  #   # friendly_name => uuid_name
+  #   "welcome.png" => "123-abc-456.png"
+  # }
+  def add_starter_asset(friendly_name, uuid_name)
```
Comment (davidsbailey) [inline dashboard/app/models/levels/level.rb]: it seems like this code (and the `starter_assets` in the `serialized_attrs` list) would better belong in dashboard/app/models/levels/applab.rb, since assets are specific to applab. They might better fit here if we thought we might generalize to support other level types, but AFAIK we aren't going to need this for gamelab or other levels. apologies for not catching this in the previous PR.

File: dashboard/test/controllers/level_starter_assets_controller_test.rb [hunk 6]
```diff
@@ -54,6 +70,72 @@ class LevelStarterAssetsControllerTest < ActionController::TestCase
...
+    post :upload, params: {level_name: level.name, files: [@file]}
+
+    level.reload
+    assert_equal 1, level.starter_assets.length
+    assert_response :success
+    summary = JSON.parse(response.body)
+    assert_equal @filename, summary['filename']
+    assert_equal 'image', summary['category']
+    assert_equal 123, summary['size']
+  end
```
Comment (davidsbailey) [inline dashboard/test/controllers/level_starter_assets_controller_test.rb:155]: great tests! please add one test case for a friendly name containing at least the `'` and `"` characters.

**Review by davidsbailey — COMMENTED** [review]
great progress, Maddie! A few more comments inline.

File: dashboard/app/models/levels/level.rb [hunk 7]
```diff
@@ -602,6 +602,19 @@ def age_13_required?
...
+    self.starter_assets ||= {}
+    self.starter_assets[friendly_name] = uuid_name
+    save
```
Comment (davidsbailey) [inline dashboard/app/models/levels/level.rb]: `save!` would be better here, so that we see the error

File: dashboard/test/controllers/level_starter_assets_controller_test.rb [hunk 8]
```diff
@@ -54,6 +70,72 @@ class LevelStarterAssetsControllerTest < ActionController::TestCase
...
+  test 'upload: returns unprocessable_entity if file fails to upload' do
+    LevelStarterAssetsController.any_instance.
+      expects(:get_object).
+      returns(@file_obj)
+    @file_obj.expects(:upload_file).returns(false)
+
+    sign_in create(:levelbuilder)
+    post :upload, params: {level_name: create(:level).name, files: [@file]}
+
+    assert_response :unprocessable_entity
```
Comment (davidsbailey) [inline dashboard/test/controllers/level_starter_assets_controller_test.rb:122]: 👍 `:unprocessable_entity` seems like a good choice here.

File: dashboard/test/controllers/level_starter_assets_controller_test.rb [hunk 9]
```diff
@@ -54,6 +70,72 @@ class LevelStarterAssetsControllerTest < ActionController::TestCase
...
+    LevelStarterAssetsController.any_instance.
+      expects(:get_object).
+      returns(@file_obj)
+    @file_obj.expects(:upload_file).returns(true)
+    Level.any_instance.expects(:save).returns(false)
+
+    sign_in create(:levelbuilder)
+    post :upload, params: {level_name: create(:level).name, files: [@file]}
+
+    assert_response :unprocessable_entity
```
Comment (davidsbailey) [inline dashboard/test/controllers/level_starter_assets_controller_test.rb]: once you are using `save!`, I think this could turn into
```ruby
assert_raises(ActiveRecord::RecordInvalid) do
  post ...
```

**Review by davidsbailey — COMMENTED** [review]
a couple more comments which came to mind while reviewing the PR after this one

File: dashboard/app/models/levels/level.rb (hunk 5 above)
Comment (maddiedierker) [reply dashboard/app/models/levels/level.rb]: that's okay! i think that's a great idea to have `starter_assets` live on Applab instead of the Level model. i'll do this in a follow-up PR

File: dashboard/config/routes.rb (hunk 2 above)
Comment (maddiedierker) [reply dashboard/config/routes.rb:237]: i think both of these routes need member entries for different but similar reasons:

- Rails convention for `:get` is `GET /level_starter_assets/:level_name`, but we need `GET /level_starter_assets/:level_name/:filename` in order to get the resource
- Rails convention for `:post` is `POST /level_starter_assets`, but we need `POST /level_starter_assets/:level_name` to create the resource

are there other/better ways to define these special cases rather than by using members?

Comment (davidsbailey) [reply dashboard/config/routes.rb:237]: something similar is possible based on the following example:
```ruby
  resources :scripts, path: '/s/' do
    # /s/xxx/stage/yyy/puzzle/zzz
    resources :stages, only: [], path: "/stage", param: 'position', format: false do
      resources :script_levels, only: [:show], path: "/puzzle", format: false do
```
that example ends up being slightly different because it uses extra path parts (`/foo/:foo_id/bar/:bar_id` instead of `/foo/:foo_id/:bar_id`). it may be possible to eliminate those, but at that point it'll probably get more complicated than what you have here.

One idea would be to use the routes:
```
POST /level_names/:level_name/starter_assets
GET /level_names/:level_name/starter_assets/:id
DELETE /level_names/:level_name/starter_assets/:id
```
which you could get by doing
```ruby
resources :level_names, only: [] do
  resources :starter_assets, only: [:create, :show, :destroy]
end
```
this is a bit more railsy but still a a bit weird because you are accessing the level by name whereas the rails way would be to access it by id.

its fine if you want to punt on this but worth noting that it'll be harder to change once external users start relying on the GET path (the POST and DELETE aren't too bad because levelbuilder updates are a bit more atomic and so you don't really get the deploy problems described in https://github.com/code-dot-org/code-dot-org/pull/30345)

Comment (maddiedierker) [reply dashboard/config/routes.rb:237]: yeah, i think i will punt on this because the end goal is to totally remove the GET path in favor of refactoring the client to rely on an S3 URL rather than endpoints on our server that read from S3

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 10]
```diff
@@ -1,8 +1,11 @@
 class LevelStarterAssetsController < ApplicationController
+  authorize_resource class: false, except: [:show, :file]
+  before_action :require_levelbuilder_mode, except: [:show, :file]
   before_action :set_level
 
   S3_BUCKET = 'cdo-v3-assets'.freeze
   S3_PREFIX = 'starter_assets/'.freeze
+  VALID_FILE_EXTENSIONS = %w(.jpg .jpeg .gif .png)
```
Comment (davidsbailey) [inline dashboard/app/controllers/level_starter_assets_controller.rb]: I'd suggest including `.mp3`, assuming that'll essentially give us the ability to upload initial sounds "for free". if that doesn't work, then let's at least make sure something sane happens in the ui when a levelbuilder tries to upload a sound file in start mode. again this can be a separate PR if needed

**Review by davidsbailey — COMMENTED** [review]
updates look good! one more comment below

File: dashboard/app/controllers/level_starter_assets_controller.rb (hunk 10 above)
Comment (maddiedierker) [reply dashboard/app/controllers/level_starter_assets_controller.rb]: good call. we wanted to scope this feature to only image assets, but uploading/rendering an .mp3 works just fine, so we might as well add that to the list of valid extensions!

**Comment by maddiedierker** [comment]
@davidsbailey thank you for all the great feedback! i think i have implemented everything, so PTAL!

**Review by davidsbailey — APPROVED** [review]
👏 👏 👏 great work Maddie!
